File: src/main/java/game/Game.java
package game;

import java.util.*;

public class Game {
    private final int maxRounds;
    private int round;
    private GameVisitor visitor;
    private final Player[] players;
    private int drafter;

    public Game(int maxRounds, PlayerConfiguration playerConfiguration) {
        this.maxRounds = maxRounds;
        players = playerConfiguration.build();
    }

    public Player[] players() {
        return players;
    }

    public void advanceDraft(Auswahl auswahl) {
        Player current = players[drafter];
        current.setAuswahl(auswahl);
        visitor.draw(current);
        drafter++;
        if (drafter >= players.length) {
            pushMatchResult();
        } else if (players[drafter].getType() == PlayerType.AI) {
            int zufallsZahl = (int) (Math.random() * 5);
            advanceDraft(Auswahl.values()[zufallsZahl]);
        }
    }

    private void resetRound() {
        drafter = 0;
        for (Player player : players) {
            player.setAuswahl(null);
        }
    }

    public void setVisitor(GameVisitor visitor) {
        this.visitor = visitor;
    }

    public void pushMatchResult() {
        Map<Player, Integer> winMap = new HashMap<>();
        int heighestScore = Integer.MIN_VALUE;
        for (Player player : players) {
            int score = 0;
            for (Player opponent : players) {
                switch (player.getAuswahl().matchResult(opponent.getAuswahl())) {
                    case Verloren -> score--;
                    case Gewonnen -> score++;
                }
            }
            if (score > heighestScore) {
                heighestScore = score;
            }
            System.out.println(player.displayName() + " has score " + score);
            winMap.put(player, score);
        }
        Collection<Player> winners = new ArrayList<>();
        Collection<Player> loosers = new ArrayList<>();
        for (Map.Entry<Player, Integer> entry : winMap.entrySet()) {
            (entry.getValue() == heighestScore ? winners : loosers).add(entry.getKey());
        }
        visitor.roundComplete(winners, loosers);
        resetRound();
        if (!loosers.isEmpty()) {
            for (Player winner : winners) {
                winner.increaseScore();
            }
        }
        for (Player player : players) {
            visitor.updateScore(player);
        }

        if (++round > maxRounds) {
            Player winner = currentWinner();
            if (winner != null) {
                visitor.end(winner);
            }
        }
    }

    private Player currentWinner() {
        Player winner = Arrays.stream(players)
            .max(Comparator.comparingInt(Player::getScore))
            .orElseThrow(IllegalStateException::new); /* unreachable */
       if (checkForMultipleWinners(winner)) {
           return null;
       }
        return winner;
    }

    private boolean checkForMultipleWinners(Player oneWinner) {
        int winners = 0;    
        for (Player player : players) {
            if (player.getScore() == oneWinner.getScore() && winners++ == 1) {
                return true;
            }
        }
        return false;
    }

    public void start() {
        for (Player player : players) {
            visitor.updateScore(player);
        }
    }
}

File: src/main/kotlin/game/Game.kt
package game

import game.PlayerConfiguration
import game.GameVisitor
import game.Player
import game.Auswahl
import game.PlayerType
import game.RoundResult
import java.util.function.ToIntFunction
import java.lang.IllegalStateException
import java.util.function.Supplier
import java.awt.image.ImageProducer
import java.awt.image.FilteredImageSource
import java.awt.image.CropImageFilter
import game.Game
import java.util.*
import java.util.function.BiConsumer

class Game(private val maxRounds: Int, playerConfiguration: PlayerConfiguration) {
    private var round = 0
    private var visitor: GameVisitor? = null
    private val players: Array<Player>
    private var drafter = 0

    init {
        players = playerConfiguration.build()
    }

    fun players(): Array<Player> {
        return players
    }

    fun advanceDraft(auswahl: Auswahl?) {
        val current = players[drafter]
        current!!.auswahl = auswahl
        visitor!!.draw(current)
        drafter++
        if (drafter >= players.size) {
            pushMatchResult()
        } else if (players[drafter]?.type == PlayerType.AI) {
            val zufallsZahl = (Math.random() * 5).toInt()
            advanceDraft(Auswahl.values()[zufallsZahl])
        }
    }

    private fun resetRound() {
        drafter = 0
        for (player in players) {
            player?.auswahl = null
        }
    }

    fun setVisitor(visitor: GameVisitor?) {
        this.visitor = visitor
    }

    fun pushMatchResult() {
        val result = RoundResult(players)
        val winners = result.winners()
        val loosers = result.loosers()
        visitor!!.roundComplete(result)
        resetRound()
        result.mayRewardPlayers()
        for (player in players!!) {
            visitor!!.updateScore(player)
        }
        if (++round > maxRounds) {
            val winner = currentWinner()
            if (winner != null) {
                visitor!!.end(winner)
            }
        }
    }

    private fun currentWinner(): Player? {
        val winner = Arrays.stream(players)
                .max(Comparator.comparingInt { obj: Player -> obj.score })
                .orElseThrow { IllegalStateException() } /* unreachable */
        return if (checkForMultipleWinners(winner)) {
            null
        } else winner
    }

    private fun checkForMultipleWinners(oneWinner: Player?): Boolean {
        var winners = 0
        for (player in players) {
            if (player.score == oneWinner?.score && winners++ == 1) {
                return true
            }
        }
        return false
    }

    fun start() {
        for (player in players!!) {
            visitor!!.updateScore(player)
        }
    }
}